File: C++/프로그래머스/0/120861. 캐릭터의 좌표/캐릭터의 좌표.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<int> solution(vector<string> keyinput, vector<int> board) {
    // 원본 board 훼손 방지
    vector<int> answer = {0, 0};
    
    const vector<int> max_pos = {
        board[1] / 2, -board[1] / 2, //top, bottom
        board[0] / 2, -board[0] / 2, //right, left
    };
    
    for (const auto k: keyinput) {
        if (k == "up" && answer[1] + 1 <= max_pos[0]) {
            answer[1]++;
            continue;
        }
        
        if (k == "down" && answer[1] - 1 >= max_pos[1]) {
            answer[1]--;
            continue;
        }
        
        if (k == "right" && answer[0] + 1 <= max_pos[2]) {
            answer[0]++;
            continue;
        }
        
        if (k == "left" && answer[0] - 1 >= max_pos[3]) {
            answer[0]--;
            continue;
        }
    }
    
    return answer;
}

File: Go/프로그래머스/0/120861. 캐릭터의 좌표/캐릭터의 좌표.go
func solution(keyinput []string, board []int) []int {
    x_max := board[0] / 2
    x_min := -board[0] / 2
    y_max := board[1] / 2
    y_min := -board[1] / 2
    
    pos := make([]int, 2)
    
    for _, key := range keyinput {
        switch {
        case key == "right" && pos[0] < x_max:
            pos[0]++
        case key == "left" && pos[0] > x_min:
            pos[0]--
        case key == "up" && pos[1] < y_max:
            pos[1]++
        case key == "down" && pos[1] > y_min:
            pos[1]--
        }
    }
    
    
    return pos
}